File: ruby/easy/subtree_of_another_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} s
# @param {TreeNode} t
# @return {Boolean}
def is_subtree(s, t)
    queue = []
    queue.push(s) if s
    while !queue.empty?
        current = queue.shift
        return true if matches(current, t)
        queue.push(current.left) if current.left
        queue.push(current.right) if current.right
    end
    return false
end

def matches(s, t)
    if(s.nil? || t.nil?)
        return (s.nil? && t.nil?)
    end
    return s.val == t.val && matches(s.left, t.left) && matches(s.right, t.right)
end

File: javascript/easy/subtreeOfAnotherTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 */
var isSubtree = function(s, t) {
    const set = new Set();
    dfs(s, set);
    const res = dfs(t, null);
    return set.has(res);
};

var dfs = function(s, set) {
    if(s === null) return '$';
    str = s.val + dfs(s.left, set) + dfs(s.right, set);
    if(set !== null) {
        set.add(str);    
    }
    return str;
};
